torch/csrc/distributed/c10d/init.cpp

```diff
@@ -992,6 +1010,36 @@ that adds a prefix to each key inserted to the store.
           "get_future",
           [](::c10d::ProcessGroup::Work& work)
               -> std::shared_ptr<jit::PythonFutureWrapper> {
+#ifdef USE_C10D_NCCL
+            auto fut = dynamic_cast<::c10d::ProcessGroupNCCL::FutureNCCL*>(
+                work.getFuture().get());
+            // By default record_stream_cb_ of the future is not set, and here
+            // we assume the type of this future value is or can be casted into
+            // a tensor vector that only has exactly one tensor.
+            if (!fut->hasRecordStreamCallback()) {
```

Comment:
Could there be race conditions where different threads could try to set this at the same time? Do we support user calling this code from multithreaded environment?

Reply:
The asynchronous threads/streams on the same device actually will be chained in the order of creation time (i.e., forked by the call of then). See the comment on addCallback function in ProcessGroupNCCL.hpp.